# The main entry point of your workflow.
# After configuring, running snakemake -n in a clone of this repository should successfully execute a dry-run of the workflow.
# Config file should have unique sample names for each sample. Rows with the
# same samples name will be merged (eg, for combining the same sample
# sequencing across multiple lanes with multiple sets of fastq)

# ============================================================================
# RNA-seq Snakemake Pipeline
# ============================================================================

configfile: "config/config.yaml"

# ============================================================================
# Modules and constraints
# ============================================================================

include: "rules/common.py"

wildcard_constraints:
    GenomeName = "|".join(STAR_genomes.index),
    sample = "|".join(AllSamples),
    Strandedness = "U|FR|RF"

# ============================================================================
# Rules
# ============================================================================

include: "rules/IndexGenome.smk"
include: "rules/PreprocessAndAlign.smk"
include: "rules/QC.smk"
include: "rules/ExpressionAnalysis.smk"
include: "rules/MakeBigwigs.smk"
include: "rules/SplicingAnalysis.smk"

# ============================================================================
# Targets
# ============================================================================

rule all:
    input:
        # Genomes
        rules.Gather_used_STAR_Indexes.input,
        
        # Alignments
        expand("results/Alignments/{sample}/Aligned.sortedByCoord.out.bam", sample=AllSamples),
        expand("results/Alignments/{sample}/Aligned.sortedByCoord.out.bam.indexing_done", sample=AllSamples),

        # Expression
        expand("results/featureCounts/{GenomeName}/AllSamplesUnstrandedCounting.Counts.txt", GenomeName = samples['STARGenomeName'].unique()),

        # Differential Expression with EdgeR
        #expand("differential_expression/{contrast}/results.tsv.gz", contrast=contrasts),
        #expand("ExpressionMatrices/{GenomeName}/log2TPM.bed", GenomeName = samples['STARGenomeName'].unique()),
        
        # QC
        "results/QC/ReadCountsPerSamples.tsv",
        "results/Multiqc",

        # Visualization with BigWigs 
        expand("results/bigwigs/unstranded/{sample}.bw", sample=AllSamples),

        # Splicing Analysis Using Leafcutter
        ## This will get junction counts and run the junction classifier from leafcutter2 
        expand("results/SplicingAnalysis/leafcutter/{GenomeName}/juncTableBeds/{Metric}.sorted.bed.gz", GenomeName = samples['STARGenomeName'].unique(), Metric = ["JuncCounts", "PSI", "PSI_ByMax", "PSIDenom", "PSI_ByMaxDenom"]),
        expand("results/SplicingAnalysis/ClassifyJuncs/{GenomeName}/Leaf2_junction_classifications.txt", GenomeName = samples['STARGenomeName'].unique()),
        ### This will add splice site scores to the splice sits ID'd with leafcutter. It does this by comparing to a PWM based on all annotated 5'/3'ss
        expand("results/SplicingAnalysis/ObservedJuncsAnnotations/{GenomeName}.uniq.annotated.with_ss_scores.tsv.gz", GenomeName = samples['STARGenomeName'].unique()),

        # Differential Splicing With Leafcutter
        expand("results/SplicingAnalysis/differential_splicing_tidy/{contrast}/Results.tsv.gz", contrast=contrasts),






        




rule Gather_Fastp_Fastqs:
    input:
        expand("FastqFastp/{sample}.fastp.html", sample=AllSamples),





        # expand("SplicingAnalysis/ObservedJuncsAnnotations/{GenomeName}.uniq.annotated.with_ss_scores.tsv.gz", GenomeName = samples['STARGenomeName'].unique()),
        # expand("SplicingAnalysis/leafcutter/{GenomeName}/juncTableBeds/{Metric}.sorted.bed.gz", GenomeName = samples['STARGenomeName'].unique(), Metric = ["JuncCounts", "PSI", "PSI_ByMax", "PSIDenom", "PSI_ByMaxDenom"]),
        # expand("SplicingAnalysis/ClassifyJuncs/{GenomeName}/Leaf2_junction_classifications.txt", GenomeName = samples['STARGenomeName'].unique()),
        # expand("SplicingAnalysis/differential_splicing_tidy/{contrast}/Results.tsv.gz", contrast=contrasts),
        # expand("differential_expression/{contrast}/results.tsv.gz", contrast=contrasts),